File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.swift
import Foundation

func solution(_ numbers:[Int], _ target:Int) -> Int {
    var answer = 0
    
    func dfs(_ number: Int, _ depth: Int) {
        if depth == numbers.count-1 {
            if number == target {
                answer += 1
            }
            return
        }
        
        dfs(number+numbers[depth+1], depth+1)
        dfs(number-numbers[depth+1], depth+1)
    }
    
    dfs(numbers.first!, 0)
    dfs(-numbers.first!, 0)
    
    return answer
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

class Solution {
    
    int answer = 0;
    int[] numbers;
    int target;
    
    public int solution(int[] numbers, int target) {
        this.numbers = numbers;
        this.target = target;
        dfs(0, 0);
        
        return answer;
    }
    
    void dfs(int value, int depth) {
        if (depth == numbers.length) {
            if (value == target) { 
                answer++; 
            }
            return;
        }
            
        dfs(value + numbers[depth], depth + 1);
        dfs(value - numbers[depth], depth + 1);
    }
}